vowpalwabbit/config/options_cli.cc

```diff
@@ -0,0 +1,365 @@
+        if (opt_value.empty()) { THROW("Argument for " << current_opt << " should follow =") }
+        m_tokens[current_opt].push_back(std::vector<std::string>{});
+        m_tokens[current_opt].back().push_back(opt_value);
+        m_supplied_options.insert(current_opt);
+      }
+    }
+    else if (arg.find('-') == 0)
+    {
+      // This is actually a negative value, treat it as a value.
+      if (is_number(arg))
+      {
+        // direct value
+        if (last_opt.empty()) { m_positional_tokens.push_back(arg); }
+        else
+        {
+          m_tokens[last_opt].back().push_back(arg);
+        }
+      }
+      else
+      {
+        auto current_opt = arg.substr(1);
+        last_opt = current_opt;
+        m_tokens[current_opt].push_back(std::vector<std::string>{});
+        m_supplied_options.insert(current_opt);
+      }
+    }
+    else
+    {
+      // direct value
+      if (last_opt.empty()) { m_positional_tokens.push_back(arg); }
+      else
+      {
+        m_tokens[last_opt].back().push_back(arg);
+      }
+    }
+  }
+}
+
+options_cli::options_cli(std::vector<std::string> args) : m_command_line(std::move(args)) { process_command_line(); }
+
+void options_cli::internal_add_and_parse(const option_group_definition& group)
+{
+  for (const auto& opt_ptr : group.m_options)
+  {
+    details::cli_typed_option_handler handler(*this);
+    opt_ptr->accept(handler);
+    m_defined_options[opt_ptr->m_name].push_back(opt_ptr);
+    if (!opt_ptr->m_short_name.empty()) { m_defined_options[opt_ptr->m_short_name].push_back(opt_ptr); }
+  }
+
+  const auto contains_necessary_options = group.contains_necessary_options();
+  const auto is_necessary_enabled = group.check_necessary_enabled(*this);
+
+  // These options are only reachable if necessary was also passed.
+  for (const auto& opt_ptr : group.m_options)
+  {
+    if ((contains_necessary_options && is_necessary_enabled) || !contains_necessary_options)
+    {
+      m_reachable_options.insert(opt_ptr->m_name);
+      if (!opt_ptr->m_short_name.empty()) { m_reachable_options.insert(opt_ptr->m_short_name); }
+    }
+
+    if (contains_necessary_options)
+    {
+      // We need to convert the unordered set to an ordered one for stable output.
+      std::set<std::string> necessary_flags_set(group.m_necessary_flags.begin(), group.m_necessary_flags.end());
+      m_dependent_necessary_options[opt_ptr->m_name].push_back(necessary_flags_set);
+      if (!opt_ptr->m_short_name.empty())
+      { m_dependent_necessary_options[opt_ptr->m_short_name].push_back(necessary_flags_set); }
+    }
+  }
+}
+
+bool options_cli::was_supplied(const std::string& key) const { return m_tokens.find(key) != m_tokens.end(); }
+
+void options_cli::check_unregistered(VW::io::logger& logger)
+{
+  for (auto const& supplied : m_supplied_options)
+  {
+    if (m_defined_options.count(supplied) == 0)
+    { THROW_EX(VW::vw_unrecognised_option_exception, "unrecognised option '--" << supplied << "'") }

```

Comment:
what if an unrecognized short name is provided?

Reply:
Good point we can detect based on length